source/common/http/codec_client.cc

```diff
@@ -117,6 +121,11 @@ void CodecClient::onReset(ActiveRequest& request, StreamResetReason reason) {
   }
 
   deleteRequest(request);
+
+  // Remember if the last request was reset
+  if (active_requests_.empty()) {
+    last_request_reset_ = true;
+  }
```

Comment:
This portion of the change makes me a little uncomfortable. Even in the case of resets, I think there might be cases in which for example we reset HTTP/2 streams and then still want to gracefully close the connection. I'm honestly not sure of the implications of SO_LINGER. I guess my question is: in the case where we don't flush (NoFlush) can we just always do SO_LINGER off in the case of source port 0? Or does that cause issues in the happy path in some of your test cases? 

@alyssawilk do you have any opinions on ^?

If we keep this part of the change this needs tests.

Reply:
Assuming by "SO_LINGER off" you mean the unix linger on with timeout 0 like used in this PR.

IMO whenever we close with NoFlush we imply that we do not care to try to send or receive any more data, even if buffered either in userspace or in kernel. So functionally SO_LINGER off would be fine, as it extends the userspace behavior (throw away any buffered data) to the kernel side (throw away any data in socket buffers).

My understanding is that with "SO_LINGER off", on client close() a TCP RST will be sent to the server. In general it would be good to avoid TCP RSTs when not needed, as they may look confusing on packet traces.

Reply:
Then should we perhaps only do this in certain kind of resets? Like a reset due to a connection network / bind error? I'm not thrilled about doing this generically like this.

Reply:
If the no-linger close should be initiated only when upstream closed its end abortively and envoy got RST, then I think this logic could go into the ConnectionImpl::onFileEvent(). I think it is called with the early close int this case. Perhaps it needs an additional flag to do it for client connections specifically.